### ISSUE TITLE
Single file component should be functional if component has no data

### ISSUE THREAD

**Issue by rightaway** [vuejs/eslint-plugin-vue#1087]
**Please describe what the rule should do:**

If a single file component doesn't have `data`, the rule should alert you to use `<template functional>` since a component without data can be a functional component, which increases performance.

**What category should the rule belong to?**

- [x] Enforces code style
- [ ] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**
```vue
<template>
  <div>template</div>
</template>

This should be `<template functional>` because there's no `data`.
```

**Additional context**

**Comment by sosukesuzuki** [comment]
I think this proposed rule would make a sense. Can I work on this?

**Comment by ota-meshi** [comment]
I think this rule makes sense as you say.
However, the `<template functional>` is a feature that will be removed in Vue3.
I'm wondering if this rule that recommending `<template functional>` should be added to this plugin 🤔

https://github.com/vuejs/rfcs/blob/master/active-rfcs/0007-functional-async-api-change.md

What do you think?

**Comment by sosukesuzuki** [comment]
I didn't know the feature will be removed...
If eslint-plugin-vue continues supporting Vue 2, we should support this rule. However, we should close this issue without adding the rule if eslint-plugin-vue intend to support only Vue 3 in the future.

**Comment by ota-meshi** [comment]
Currently the maintainers of this plugin are not active except me, so if this status persists, it's difficult to add and maintain rules related to features that will be removed in Vue3.

I'm not going to stop supporting Vue2, but I think I'll probably just spend time maintaining Vue3 related rules once Vue3 is released and becomes available to users.

Also, by making sure that users don't use features that will be removed in Vue3 will ease the transition to Vue3. So I think that adding rules that recommend features that will be removed has a low priority.

**Comment by sosukesuzuki** [comment]
Thank you for your response. I agree with you.
